src/api/controller/api/publish.js

```diff
@@ -1,25 +1,82 @@
+import omit from 'lodash.omit';
+import Koa from 'koa';
+import route from 'koa-route';
+
 /* eslint no-await-in-loop: off */
 import getDocumentTransformer from '../../../common/getDocumentTransformer';
 
-export const publishMiddleware = async (ctx) => {
-    const count = await ctx.dataset.count({});
+const app = new Koa();
+
+export const tranformAllDocument = async (count, findLimitFromSkip, insertBatch, transformer) => {
     let handled = 0;
+    while (handled < count) {
+        const dataset = await findLimitFromSkip(100, handled);
+        const transformedDataset = await Promise.all(dataset.map(transformer));
+        await insertBatch(transformedDataset);
+        handled += dataset.length;
+    }
+};
+
+export const addTransformResultToDoc = transform => async doc => ({
+    ...omit(doc, ['_id']),
+    ...await transform(doc),
+});
+
+export const addUriToTransformResult = transformDocument => async doc => ({
+    ...await transformDocument(doc),
+    uri: doc.uri,
+});
+
+export const doPublish = async (ctx) => {
+    const count = await ctx.dataset.count({});
 
     const columns = await ctx.field.findAll();
 
-    const transformDocument = ctx.getDocumentTransformer({ env: 'node' }, columns);
+    const uriIndex = columns.findIndex(col => col.name === 'uri');
+    const getUri = ctx.getDocumentTransformer({ env: 'node', dataset: ctx.dataset }, [columns[uriIndex]]);
+    const addUri = ctx.addTransformResultToDoc(getUri);
 
-    while (handled < count) {
-        const dataset = await ctx.dataset.findLimitFromSkip(100, handled);
-        const transformedDataset = await Promise.all(dataset.map(transformDocument));
-        await ctx.publishedDataset.insertMany(transformedDataset);
-        handled += dataset.length;
-    }
+    await ctx.tranformAllDocument(
+        count,
+        ctx.dataset.findLimitFromSkip,
+        ctx.uriDataset.insertBatch,
+        addUri,
+    );
+
+    const transformDocument = ctx
+        .getDocumentTransformer({
+            env: 'node',
+            dataset: ctx.uriDataset,
+        }, columns.filter((_, index) => index !== uriIndex));
+
+    const transformDocumentAndKeepUri = ctx.addUriToTransformResult(transformDocument);
+
+    await ctx.tranformAllDocument(
+        count,
+        ctx.uriDataset.findLimitFromSkip,
+        ctx.publishedDataset.insertBatch,
+        transformDocumentAndKeepUri,
+    );
 
     ctx.redirect('/api/publication');
 };
 
-export default async (ctx) => {
+export const tryPublish = async (ctx, next) => {
```

Comment:
`tryPublish` ? I expected a transformation to occur here. Isn't an initializer for the post middleware ?